File: archiso/airootfs/usr/share/churros/services/power.py
import os
import subprocess


def _has_swap():

    try:

        out = subprocess.run(

            ["swapon", "--show", "--noheadings"],

            capture_output=True,
            text=True,
            timeout=2

        ).stdout.strip()

        return bool(out)

    except Exception:

        return False


def _current_desktop():

    return (

        os.environ.get("XDG_CURRENT_DESKTOP")

        or os.environ.get("XDG_SESSION_DESKTOP")

        or os.environ.get("DESKTOP_SESSION")

        or ""

    ).lower()


class PowerService:

    @staticmethod
    def lock():

        try:

            subprocess.Popen(
                ["loginctl", "lock-session"]
            )

        except Exception:
            pass

    @staticmethod
    def logout():

        desktop = _current_desktop()

        try:

            if "niri" in desktop:

                subprocess.Popen(["niri", "msg", "action", "quit"])

            elif "hyprland" in desktop:

                subprocess.Popen(["hyprctl", "dispatch", "exit"])

            elif "sway" in desktop:

                subprocess.Popen(["swaymsg", "exit"])

            else:

                subprocess.Popen(["loginctl", "terminate-user", str(os.getuid())])

        except Exception:
            pass

    @staticmethod
    def suspend():

        try:

            subprocess.Popen(["systemctl", "suspend"])

        except Exception:
            pass

    @staticmethod
    def can_hibernate():

        return _has_swap()

    @staticmethod
    def hibernate():

        try:

            subprocess.Popen(["systemctl", "hibernate"])

        except Exception:
            pass

    @staticmethod
    def restart():

        try:

            subprocess.Popen(["systemctl", "reboot"])

        except Exception:
            pass

    @staticmethod
    def shutdown():

        try:

            subprocess.Popen(["systemctl", "poweroff"])

        except Exception:
            pass

File: rust/services/src/power.rs
// Servicio de energía/sesión (equivalente a services/power.py)

use crate::{run, spawn};

fn has_swap() -> bool {
    match run(&["swapon", "--show", "--noheadings"], 2000) {
        Some((_, out, _)) => !out.trim().is_empty(),
        None => false,
    }
}

fn current_desktop() -> String {
    std::env::var("XDG_CURRENT_DESKTOP")
        .or_else(|_| std::env::var("XDG_SESSION_DESKTOP"))
        .or_else(|_| std::env::var("DESKTOP_SESSION"))
        .unwrap_or_default()
        .to_lowercase()
}

fn current_uid() -> String {
    run(&["id", "-u"], 2000)
        .map(|(_, out, _)| out.trim().to_string())
        .unwrap_or_default()
}

pub fn lock() {
    spawn(&["loginctl", "lock-session"]);
}

pub fn logout() {
    let desktop = current_desktop();
    if desktop.contains("niri") {
        spawn(&["niri", "msg", "action", "quit"]);
    } else if desktop.contains("hyprland") {
        spawn(&["hyprctl", "dispatch", "exit"]);
    } else if desktop.contains("sway") {
        spawn(&["swaymsg", "exit"]);
    } else {
        let uid = current_uid();
        if !uid.is_empty() {
            spawn(&["loginctl", "terminate-user", &uid]);
        }
    }
}

pub fn suspend() {
    spawn(&["systemctl", "suspend"]);
}

pub fn can_hibernate() -> bool {
    has_swap()
}

pub fn hibernate() {
    spawn(&["systemctl", "hibernate"]);
}

pub fn restart() {
    spawn(&["systemctl", "reboot"]);
}

pub fn shutdown() {
    spawn(&["systemctl", "poweroff"]);
}
